### PR TITLE
Add Segmented sort

### LINKED ISSUE #4603: [FEA] Segmented Sorting
**Is your feature request related to a problem? Please describe.**

Given a table and a set of offsets that demarcate "segments" of the table, I would like to be able to sort each segment.

**Describe the solution you'd like**

Implement a `segmented_sort` API like:
```
unique_ptr<table> segmented_sort(table_view table_to_sort, vector<size_type> key_columns, vector<size_type> segment_offsets, vector<order> column_order, vector<null_order> null_orders);
```

**Describe alternatives you've considered**
Alternatively, you can use `slice` to get a set of views and sort each view separately, but that's inefficient.

**Additional context**
I thought of this while testing `partition` in https://github.com/rapidsai/cudf/pull/4472. Since the ordering per partition is non-deterministic, it is difficult to unit test when there are multiple values per partition. 

Unfortunately, CUB's segmented sort can't be used because we need a comparison sort for the lexicographic row ordering.

Instead, a straightforward way to implement this:
1. Prepend an integral dummy column of 0s to `table_to_sort`
2. Scatter 1s to every location indicated by `segment_offsets`
3. Compute exclusive_scan of dummy column
4. Do a lexicographic sort of dummy column + `table_to_sort`

### PR DESCRIPTION
**Pull request by karthikeyann**
addresses part of #6541 Segment sort of lists

- [x] lists_column_view segmented_sort
- [x] numerical types (cub segmented sort limitation)
- [x] sort_lists(table_view)
- [x] unit tests

closes  #4603 Segmented sort
- [x] segmented_sort
- [x] unit tests.

### CODE REVIEW

File: cpp/include/cudf/lists/sorting.hpp [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+ *
+ * `source_column` with depth 1 is only supported. This uses segmented radix sort.
+ *
+ * * @code{.pseudo}
+ * source_column            : [{4, 2, 3, 1}, {1, 2, NULL, 4}, {-10, 10, 0}]
+ *
+ * Ascending,  Null After   : [{1, 2, 3, 4}, {1, 2, 4, NULL}, {-10, 0, 10}]
+ * Ascending,  Null Before  : [{1, 2, 3, 4}, {NULL, 1, 2, 4}, {-10, 0, 10}]
+ * Descending, Null After   : [{4, 3, 2, 1}, {4, 2, 1, NULL}, {10, 0, -10}]
+ * Descending, Null Before  : [{4, 3, 2, 1}, {NULL, 4, 2, 1}, {10, 0, -10}]
```
Comment (jrhemstad) [inline cpp/include/cudf/lists/sorting.hpp:41]: This isn't correct. Null After implies nulls are _larger_ than all other values, null before implies nulls are _smaller_ than all other values.

File: cpp/include/cudf/lists/sorting.hpp [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+/**
+ * @addtogroup lists_sort
+ * @{
+ * @file
+ */
+
+/**
+ * @brief Segmented sort of the elements within a list in each row of a list column.
+ *
+ * `source_column` with depth 1 is only supported. This uses segmented radix sort.
```
Comment (jrhemstad) [inline cpp/include/cudf/lists/sorting.hpp:33]: This is an implementation detail that need not be part of the documentation. 
```suggestion
 * `source_column` with depth 1 is only supported. 
```

File: cpp/include/cudf/lists/sorting.hpp [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+ * @endcode
+ *
+ * @param source_column View into the list column of numeric types to gather from
+ * @param column_order The desired sort order
+ * @param null_precedence The desired order of null compared to other elements in the list
+ * @param mr Device memory resource to allocate any returned objects
+ * @return list column with elements in each list sorted.
+ *
+ */
+std::unique_ptr<column> segmented_sort(
```
Comment (jrhemstad) [inline cpp/include/cudf/lists/sorting.hpp:51]: I don't think this is an appropriate name for this functionality. The fact that it is a segmented sort is an implementation detail. Really this function is sorting each list independently. I think `sort_lists` would be a better name.

File: cpp/src/lists/segmented_sort.cu [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+                          ValueT* values_out,
+                          int num_items,
+                          int num_segments,
+                          OffsetIteratorT begin_offsets,
+                          OffsetIteratorT end_offsets,
+                          rmm::cuda_stream_view stream)
+  {
+    rmm::device_buffer d_temp_storage;
+    size_t temp_storage_bytes = 0;
+    cub::DeviceSegmentedRadixSort::SortPairs(d_temp_storage.data(),
```
Comment (jrhemstad) [inline cpp/src/lists/segmented_sort.cu:55]: Why can't we use `thrust::sort_by_key` instead?

File: cpp/src/sort/segmented_sort.cu [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+std::unique_ptr<column> segmented_sorted_order(table_view const& keys,
+                                               std::vector<order> const& column_order,
+                                               std::vector<null_order> const& null_precedence,
+                                               rmm::cuda_stream_view stream,
+                                               rmm::mr::device_memory_resource* mr)
+{
+  CUDF_EXPECTS(std::all_of(keys.begin(),
+                           keys.end(),
+                           [](column_view const& col) { return col.type().id() == type_id::LIST; }),
+               "segmented_sort only supports lists columns");
```
Comment (jrhemstad) [inline cpp/src/sort/segmented_sort.cu:71]: Why? I'd think I'd should be able to do a segmented sort of a normal column. In fact, I filed an issue asking for exactly that ages ago: https://github.com/rapidsai/cudf/issues/4603

**Review by jrhemstad — CHANGES_REQUESTED** [review]

File: cpp/src/sort/segmented_sort.cu (hunk 5 above)
Comment (karthikeyann) [reply cpp/src/sort/segmented_sort.cu:71]: #6541 requested for list columns.  This will be renamed to `sort_lists` and this check will remain.
`segmented_sort` of normal column/table will be exposed as per #4603 specifications.

File: cpp/src/lists/segmented_sort.cu (hunk 4 above)
Comment (karthikeyann) [reply cpp/src/lists/segmented_sort.cu:55]: This is for single list column - faster than `thrust::sort_by_key`. Also with `DeviceSegmentedRadixSort`, there is room for improvement with multiple streams.

Comment (jrhemstad) [reply cpp/src/lists/segmented_sort.cu:55]: Doesn't `thrust::sort_by_key` trivially map to `DeviceSegmentedRadixSort` when doing a numeric sort? Similar to how `thrust::sort` maps to `DeviceRadixSort`?

Comment (karthikeyann) [reply cpp/src/lists/segmented_sort.cu:55]: It doesn't. `thrust::sort_by_key` is not a segmented sort. `thrust::sort_by_key` sorts `values` w.r.t to `key` sorted order.

File: cpp/include/cudf/lists/sorting.hpp (hunk 1 above)
Comment (karthikeyann) [reply cpp/include/cudf/lists/sorting.hpp]: fixed

File: cpp/include/cudf/lists/sorting.hpp (hunk 3 above)
Comment (karthikeyann) [reply cpp/include/cudf/lists/sorting.hpp]: updated.

**Review by ajschmidt8 — APPROVED** [review]

File: cpp/src/lists/segmented_sort.cu [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+#include <cudf/lists/sorting.hpp>
+#include <cudf/utilities/error.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+
+#include <thrust/copy.h>
+#include <cub/device/device_segmented_radix_sort.cuh>
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_buffer.hpp>
+#include <rmm/device_uvector.hpp>
+#include <rmm/exec_policy.hpp>
```
Comment (codereport) [inline cpp/src/lists/segmented_sort.cu:35]:
```suggestion
#include <rmm/cuda_stream_view.hpp>
#include <rmm/device_buffer.hpp>
#include <rmm/device_uvector.hpp>
#include <rmm/exec_policy.hpp>

#include <thrust/copy.h>

#include <cub/device/device_segmented_radix_sort.cuh>
```

File: cpp/src/lists/segmented_sort.cu [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+  {
+    auto output =
+      cudf::detail::allocate_like(child, child.size(), mask_allocation_policy::NEVER, stream, mr);
+    mutable_column_view mutable_output_view = output->mutable_view();
+
+    auto keys = [&]() {
+      if (child.nullable()) {
+        rmm::device_uvector<T> keys(child.size(), stream);
+        auto null_replace_T = null_precedence == null_order::AFTER ? std::numeric_limits<T>::max()
+                                                                   : std::numeric_limits<T>::min();
```
Comment (codereport) [inline cpp/src/lists/segmented_sort.cu:155]:
```suggestion
        auto const null_replace_T = null_precedence == null_order::AFTER ? std::numeric_limits<T>::max()
                                                                         : std::numeric_limits<T>::min();
```

File: cpp/src/lists/segmented_sort.cu [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+                          mutable_output_view.begin<T>(),
+                          mutable_indices_view.begin<size_type>(),
+                          mutable_indices_view.begin<size_type>(),
+                          child.size(),
+                          offsets.size() - 1,
+                          offsets.begin<size_type>(),
+                          offsets.begin<size_type>() + 1,
+                          stream);
+    std::vector<std::unique_ptr<column>> output_cols;
+    output_cols.push_back(std::move(output));
```
Comment (codereport) [inline cpp/src/lists/segmented_sort.cu:194]: Is this possible?
```suggestion
    std::vector<std::unique_ptr<column>> output_cols{std::move(output)};
```

File: cpp/src/lists/segmented_sort.cu [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+  }
+};
+
+std::unique_ptr<column> sort_lists(lists_column_view const& input,
+                                   order column_order,
+                                   null_order null_precedence,
+                                   rmm::cuda_stream_view stream,
+                                   rmm::mr::device_memory_resource* mr)
+{
+  if (input.size() == 0) return {};
```
Comment (codereport) [inline cpp/src/lists/segmented_sort.cu:212]:
```suggestion
  if (input.is_empty()) return {};
```

File: cpp/src/sort/segmented_sort.cu [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_buffer.hpp>
+#include <rmm/device_uvector.hpp>
+#include <rmm/exec_policy.hpp>
+
+//#include <cub/device/device_segmented_radix_sort.cuh>
+#include <thrust/binary_search.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/iterator/transform_iterator.h>
+#include <thrust/logical.h>
```
Comment (codereport) [inline cpp/src/sort/segmented_sort.cu:39]:
```suggestion
#include <rmm/cuda_stream_view.hpp>
#include <rmm/device_buffer.hpp>
#include <rmm/device_uvector.hpp>
#include <rmm/exec_policy.hpp>

#include <thrust/binary_search.h>
#include <thrust/iterator/counting_iterator.h>
#include <thrust/iterator/transform_iterator.h>
#include <thrust/logical.h>

#include <algorithm>
#include <iterator>
#include <memory>
```

File: cpp/src/sort/segmented_sort.cu [hunk 11]
```diff
@@ -0,0 +1,228 @@
...
+  thrust::lower_bound(rmm::exec_policy(stream),
+                      offsets_minus_one + 1,
+                      offsets_minus_one + offsets.size(),
+                      counting_iter,
+                      counting_iter + segment_ids.size(),
+                      segment_ids.begin());
+  return std::move(segment_ids);
+}
+
+// same as count_elements(list). TODO: DRY.
```
Comment (codereport) [inline cpp/src/sort/segmented_sort.cu:64]: Intentionally leaving this in?

File: cpp/tests/lists/sort_lists_tests.cpp [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+#include <cudf/fixed_point/fixed_point.hpp>
+#include <cudf/lists/lists_column_view.hpp>
+#include <cudf/table/table_view.hpp>
+#include <cudf/types.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+
+#include <type_traits>
+#include <vector>
+
+#include <cudf/lists/sorting.hpp>
```
Comment (codereport) [inline cpp/tests/lists/sort_lists_tests.cpp:34]:
```suggestion
#include <cudf/utilities/type_dispatcher.hpp>
#include <cudf/lists/sorting.hpp>

#include <type_traits>
#include <vector>
```

**Review by codereport — CHANGES_REQUESTED** [review]
looks good, just a few small things

File: cpp/src/sort/segmented_sort.cu (hunk 11 above)
Comment (karthikeyann) [reply cpp/src/sort/segmented_sort.cu:65]: @davidwendt can this code be moved to a .cuh file and reuse it in `count_elements.cu`?

File: cpp/src/sort/segmented_sort.cu (hunk 5 above)
Comment (karthikeyann) [reply cpp/src/sort/segmented_sort.cu]: implemented `segmented_sort` as well.

File: cpp/src/sort/segmented_sort.cu (hunk 11 above)
Comment (davidwendt) [reply cpp/src/sort/segmented_sort.cu:65]: Certainly. My suggestion is to move into the existing `list_device_view.cuh`
Seems like we should be able to factor the logic into there sometime in the future.

Comment (karthikeyann) [reply cpp/src/sort/segmented_sort.cu]: moved.

**Review by kkraus14 — APPROVED** [review]
cmake approval

File: cpp/include/cudf/lists/list_device_view.cuh [hunk 13]
```diff
@@ -190,4 +190,19 @@ class list_device_view {
   };
 };
 
+/**
+ * @brief returns size of the list by row index
+ *
+ */
+struct list_size_functor {
+  column_device_view const d_column;
```
Comment (jrhemstad) [inline cpp/include/cudf/lists/list_device_view.cuh:198]: Seems like this should have a constructor that verifies the column is a list column.

File: cpp/src/lists/segmented_sort.cu [hunk 14]
```diff
@@ -0,0 +1,249 @@
...
+  }
+};
+
+std::unique_ptr<column> sort_lists(lists_column_view const& input,
+                                   order column_order,
+                                   null_order null_precedence,
+                                   rmm::cuda_stream_view stream,
+                                   rmm::mr::device_memory_resource* mr)
+{
+  if (input.is_empty()) return {};
```
Comment (jrhemstad) [inline cpp/src/lists/segmented_sort.cu:215]: This will return a null `unique_ptr`. Should use `empty_like`.

File: cpp/src/sort/segmented_sort.cu [hunk 15]
```diff
@@ -0,0 +1,233 @@
...
+
+  // segment offsets from first list column
+  auto lc              = lists_column_view{keys.column(0)};
+  auto offset          = lc.offsets();
+  auto segment_offsets = column_view(offset.type(),
+                                     offset.size(),
+                                     offset.head(),
+                                     offset.null_mask(),
+                                     offset.null_count(),
+                                     offset.offset() + lc.offset());
```
Comment (jrhemstad) [inline cpp/src/sort/segmented_sort.cu:175]: This looks odd. Is this just a `slice`?

File: cpp/include/cudf/sorting.hpp [hunk 16]
```diff
@@ -184,5 +184,63 @@ std::unique_ptr<column> rank(
...
+ * elements for each column in `keys`. Size must be equal to
+ * `keys.num_columns()` or empty. If empty, all columns will be sorted with
+ * `null_order::BEFORE`.
+ * @param mr Device memory resource to allocate any returned objects
+ * @return table with elements in each segment sorted.
+ *
+ */
+std::unique_ptr<table> segmented_sort(
+  table_view const& values,
+  table_view const& keys,
```
Comment (jrhemstad) [inline cpp/include/cudf/sorting.hpp:211]: The presence of keys and values implies this is a `segmented_sort_by_key`, yes?

File: cpp/include/cudf/sorting.hpp [hunk 17]
```diff
@@ -184,5 +184,63 @@ std::unique_ptr<column> rank(
...
+  column_view const& segment_offsets,
+  std::vector<order> const& column_order         = {},
+  std::vector<null_order> const& null_precedence = {},
+  rmm::mr::device_memory_resource* mr            = rmm::mr::get_current_device_resource());
+
+/**
+ * @brief Performs a lexicographic segmented sort of the list in each row of a table of list columns
```
Comment (jrhemstad) [inline cpp/include/cudf/sorting.hpp:218]: I don't understand this function. How do you do a lexicographic sort of a `table` of list columns?

**Review by jrhemstad — CHANGES_REQUESTED** [review]

File: cpp/include/cudf/sorting.hpp (hunk 16 above)
Comment (karthikeyann) [reply cpp/include/cudf/sorting.hpp:211]: Yes. skipped it because it was lengthy (I see that now it's a bad reason). I didn't really want to create 2 APIs (with and without suffix `_by_key`).

File: cpp/include/cudf/sorting.hpp (hunk 17 above)
Comment (karthikeyann) [reply cpp/include/cudf/sorting.hpp:218]: Is this clear?
`Performs a lexicographic sort of each list in the table.`

Comment (jrhemstad) [reply cpp/include/cudf/sorting.hpp:218]: So is it the same as calling sort independently on each column?

File: cpp/src/sort/segmented_sort.cu (hunk 15 above)
Comment (karthikeyann) [reply cpp/src/sort/segmented_sort.cu:175]: Yes. will replace it.

File: cpp/src/lists/segmented_sort.cu (hunk 8 above)
Comment (codereport) [reply cpp/src/lists/segmented_sort.cu:197]: This was resolved but with no comment. So not possible?

File: cpp/include/cudf/sorting.hpp (hunk 17 above)
Comment (karthikeyann) [reply cpp/include/cudf/sorting.hpp:218]: no. It will be similar to calling sort on each row of the table. Each row is a list.
The relative row order in the table remains same. In each row, the elements in list are sorted.

File: cpp/src/lists/segmented_sort.cu (hunk 8 above)
Comment (karthikeyann) [reply cpp/src/lists/segmented_sort.cu:197]: I committed it via github web. (which automatically resolves this) then it didn't compile. (reason: deleted function. Can't move initializer list of unique_ptr). I rolled back this change.

File: cpp/include/cudf/lists/list_device_view.cuh (hunk 13 above)
Comment (karthikeyann) [reply cpp/include/cudf/lists/list_device_view.cuh:198]: It's used in host and also device codes. `#if defined(__CUDA_ARCH__)` is added to verify this.

File: cpp/include/cudf/sorting.hpp (hunk 16 above)
Comment (karthikeyann) [reply cpp/include/cudf/sorting.hpp]: updated.

File: cpp/src/lists/segmented_sort.cu (hunk 14 above)
Comment (karthikeyann) [reply cpp/src/lists/segmented_sort.cu]: updated.

File: cpp/include/cudf/lists/list_device_view.cuh (hunk 13 above)
Comment (karthikeyann) [reply cpp/include/cudf/lists/list_device_view.cuh:198]: added constructor.

File: cpp/src/sort/segmented_sort.cu [hunk 18]
```diff
@@ -0,0 +1,233 @@
...
+    column_view(data_type(type_to_id<size_type>()), segment_ids.size(), segment_ids.data()));
+  keys_with_segid.insert(keys_with_segid.end(), keys.begin(), keys.end());
+  auto segid_keys = table_view(keys_with_segid);
+
+  std::vector<order> child_column_order(column_order);
+  if (not column_order.empty())
+    child_column_order.insert(child_column_order.begin(), order::ASCENDING);
+  std::vector<null_order> child_null_precedence(null_precedence);
+  if (not null_precedence.empty())
+    child_null_precedence.insert(child_null_precedence.begin(), null_order::AFTER);
```
Comment (codereport) [inline cpp/src/sort/segmented_sort.cu:116]: This is ITM, but I can't think of a better way to do it. Also, `insert` to front is linear which technically could be avoided - but once again - I can't think of a cleaner way to do it.

File: cpp/include/cudf/sorting.hpp (hunk 17 above)
Comment (jrhemstad) [reply cpp/include/cudf/sorting.hpp]: So it's the same as calling `sort_lists` on each column in the table?

File: cpp/src/sort/segmented_sort.cu [hunk 19]
```diff
@@ -0,0 +1,233 @@
...
+                                  std::vector<null_order> const& null_precedence,
+                                  rmm::cuda_stream_view stream,
+                                  rmm::mr::device_memory_resource* mr)
+{
+  CUDF_EXPECTS(keys.num_columns() > 0, "keys table should have atleast one list column");
+  std::vector<column_view> key_value_columns;
+  key_value_columns.reserve(keys.num_columns() + values.num_columns());
+  key_value_columns.insert(key_value_columns.end(), keys.begin(), keys.end());
+  key_value_columns.insert(key_value_columns.end(), values.begin(), values.end());
+  validate_list_columns(table_view{key_value_columns}, stream);
```
Comment (codereport) [inline cpp/src/sort/segmented_sort.cu:160]: My first thought was ITM, but then I also noticed that `key_value_columns` doesn't get used anywhere except line 159 in `validate_list_columns`. Seems like we should have a function `valid_key_value_columns` or something like that?

File: cpp/src/sort/segmented_sort.cu [hunk 20]
```diff
@@ -0,0 +1,233 @@
...
+               "segmented_sort only supports lists columns");
+  // check if all list sizes are equal.
+  auto table_device  = table_device_view::create(keys, stream);
+  auto counting_iter = thrust::make_counting_iterator<size_type>(0);
+  CUDF_EXPECTS(
+    thrust::all_of(rmm::exec_policy(stream),
+                   counting_iter,
+                   counting_iter + keys.num_rows(),
+                   [d_keys = *table_device] __device__(size_type idx) {
+                     auto size = list_size_functor{d_keys.column(0)}(idx);
```
Comment (codereport) [inline cpp/src/sort/segmented_sort.cu:82]:
```suggestion
                     auto const size = list_size_functor{d_keys.column(0)}(idx);
```

**Review by codereport — CHANGES_REQUESTED** [review]

File: cpp/include/cudf/sorting.hpp (hunk 17 above)
Comment (karthikeyann) [reply cpp/include/cudf/sorting.hpp]: no. it's similar to treating each row as table, then sort it. then repeat it for all rows.

Copying the example from #6541

| row# | a | b | c |
| - | - | - | - |
| 0 | [21, 22, 23, 22] | [13, 14, 12, 11] | ["a", "b", "c", "d"]
| 1 | [22, 21, 23, 22] | [14, 13, 12, 11] | ["a", "b", "c", "d"]

Here [...] is a list.


| row# | a | b | c |
| - | - | - | - |
| 0 | <ul><li>21</li><li>22</li><li>23</li><li>22</li></ul> | <ul><li>13</li><li>14</li><li>12</li><li>11</li></ul> | <ul><li>"a"</li><li>"b"</li><li>"c"</li><li>"d"</li></ul> |
| 1 | <ul><li>22</li><li>21</li><li>23</li><li>22</li></ul>| <ul><li>14</li><li>13</li><li>12</li><li>11</li></ul> | <ul><li>"a"</li><li>"b"</li><li>"c"</li><li>"d"</li></ul> |

`sort_lists(values={a,b,c}, keys={a,b})` output will be

| row# | a | b | c |
| - | - | - | - |
| 0 | <ul><li>21</li><li>22</li><li>22</li><li>23</li></ul> | <ul><li>13</li><li>11</li><li>14</li><li>12</li></ul> | <ul><li>"a"</li><li>"d"</li><li>"b"</li><li>"c"</li></ul> |
| 1 | <ul><li>21</li><li>22</li><li>22</li><li>23</li></ul>| <ul><li>13</li><li>11</li><li>14</li><li>12</li></ul> | <ul><li>"b"</li><li>"d"</li><li>"a"</li><li>"c"</li></ul> |

Comment (jrhemstad) [reply cpp/include/cudf/sorting.hpp]: Does that require all of the lists in a given row to have the same number of elements? What would the result be for this input?

| row# | a | b | c |
| - | - | - | - |
| 0 | [21, 22, 23, 22] | [13, 14] | ["a", "c", "d"]
| 1 | [22, 21, 23, 22] | [14] | ["c", "d"]

File: cpp/src/sort/segmented_sort.cu (hunk 18 above)
Comment (karthikeyann) [reply cpp/src/sort/segmented_sort.cu:116]: not clean but efficient code (avoids extra linear copy).
```suggestion
  auto prepend_default = [](auto const& vector, auto default_value) {
    if (vector.empty()) return vector;
    std::remove_cv_t<std::remove_reference_t<decltype(vector)>> pre_vector;
    pre_vector.reserve(pre_vector.size() + 1);
    pre_vector.push_back(default_value);
    pre_vector.insert(pre_vector.end(), vector.begin(), vector.end());
    return pre_vector;
  };
  auto child_column_order    = prepend_default(column_order, order::ASCENDING);
  auto child_null_precedence = prepend_default(null_precedence, null_order::AFTER);
```

File: cpp/src/sort/segmented_sort.cu (hunk 19 above)
Comment (karthikeyann) [reply cpp/src/sort/segmented_sort.cu:160]: making them as table_view has 2 benefits. 
1. checks key and values columns have same number of rows. (during table_view creation)
2. use parallelism across combined key+value columns.
moved it to `valid_key_value_list_columns`. (it does not sound generic enough like `validate_list_columns`)

File: cpp/src/sort/segmented_sort.cu (hunk 20 above)
Comment (karthikeyann) [reply cpp/src/sort/segmented_sort.cu]: updated.

File: cpp/src/sort/segmented_sort.cu (hunk 19 above)
Comment (karthikeyann) [reply cpp/src/sort/segmented_sort.cu]: updated.

**Comment by karthikeyann** [comment]
Unrelated style check failures.
`FAILED: mypy style check`
`python/cudf/cudf/core/column/categorical.py:818: error: Item "dtype[Any]" of "Union[Any, dtype[Any]]" has no attribute "serialize"`

**Comment by kkraus14** [comment]
> Unrelated style check failures.
> `FAILED: mypy style check`
> `python/cudf/cudf/core/column/categorical.py:818: error: Item "dtype[Any]" of "Union[Any, dtype[Any]]" has no attribute "serialize"`

Being fixed in https://github.com/rapidsai/cudf/pull/7279

File: cpp/include/cudf/sorting.hpp (hunk 17 above)
Comment (karthikeyann) [reply cpp/include/cudf/sorting.hpp]: Yes. 
For this input, it will throw a logic error.

**Review by jrhemstad — CHANGES_REQUESTED** [review]
High level questions:
- Why is `sort_lists` different from `segmented_sort`? Couldn't `sort_lists` just be implemented by calling `segmented_sort` on the data/offsets children of a list column?
- This brings up the question, do we even need a `sort_lists` API? Couldn't a caller trivially just do `segmented_sort(list.data, list.offsets)`? This had the advantage of implicitly enforcing the "lists of depth 1" requirement. 
- The `sort_lists` table API still sticks out as very odd to me. I think from https://github.com/rapidsai/cudf/issues/6541 what @kkraus14 really wanted is just a table level segmented sort, not a lexicographic sort of lists of all the same length. 

In summary, all I would think we need is:
- `segmented_sort_by_key(table_view values, table_view keys, column_view segment_offsets, ...)`
- `segmented_sorted_order(table_view keys, column_view segment_offsets)`

**Comment by karthikeyann** [comment]
Had discussion with @jrhemstad 
- `sort_lists` has 2 functions/implementations.
  - `sort_lists(column_view)` for single column (faster method, uses cub segmented radix sort, fixed_width columns)
  - `sort_lists(table_view)` (this is implemented by calling `segmented_sort_by_key(table_view)` on the data/offsets children of a list column)
- Caller can call `segmented_sort_by_key`, but caller needs to add offset to list.offsets, have to reconstruct the list columns of table after getting result, which is what `sort_lists(table_view)` does.
-  discussed with @jrhemstad 
  > functions for segmented sorts and sorted_order (argsort) to sort the list elements within a column

   `sort_lists(table_view)` is not needed and will be removed. `sort_lists(column_view)` is required and will be kept.

**Comment by harrism** [comment]
@karthikeyann can you get the requested changes done by tomorrow? If not we should push to 0.19.

**Comment by karthikeyann** [comment]
> @karthikeyann can you get the requested changes done by tomorrow? If not we should push to 0.19.

Yes. It will be done by today. By tomorrow, I will address re-review comments as well.

**Review by jrhemstad — APPROVED** [review]

**Review by codereport — APPROVED** [review]
:+1:

**Comment by kkraus14** [comment]
rerun tests

**Comment by kkraus14** [comment]
@gpucibot merge
